src/marker.rs

```diff
@@ -140,6 +140,42 @@ use std::os::raw::c_int;
 /// the GIL is not held.
 ///
 /// See the [module-level documentation](self) for more information.
+///
+/// # Examples
+///
+/// This tracking is currently imprecise as it relies on the [`Send`] auto trait on stable Rust.
+/// For example, an `Rc` smart pointer should be usable without the GIL, but we currently prevent that:
+///
+/// ```compile_fail
+/// # use pyo3::prelude::*;
+/// Python::with_gil(|py| {
+///     let rc = Rc::new(42);
+///
+///     py.allow_threads(|| {
+///         println!("{:?}", rc);
+///     });
+/// });
+/// ```
+///
+/// This also implies that one can circumvent this protection using e.g. the [`send_wrapper`](https://docs.rs/send_wrapper/) crate:
```

Comment:
Maybe it should be made clearer here that doing this is not sound and the fix is available on nightly Rust. Possibly also link to #2141?

Actually, looking at https://github.com/PyO3/pyo3/issues/2141#issuecomment-1028921883 again I am now wondering if the time is right to make some changes to `allow_threads` to make it sound even in the presence of `send_wrapper`?

We've done the removal of `GilGuard` now so we can potentially offer a safe-but-awkward API and then an `unsafe` but neater one? We can then document the lack of auto traits as hampering us (or maybe someone comes up with a better idea how to model this conveniently + safely).

Reply:
> Maybe it should be made clearer here that doing this is not sound and the fix is available on nightly Rust. Possibly also link to https://github.com/PyO3/pyo3/issues/2141?

Will amend the comment.

W.r.t. `allow_threads`, if you are referring to adding the `'static` bound then I think your observation that 

> I don't think 'static is very practical, because then the closure can't capture any variables except owned values by move. I think that'll turn out quite awkward.

still applies. For example, this would make `allow_threads` not work for rust-numpy when a `PyArray` is borrowed as an `ArrayView<'_>` which we could not safely work on without keeping the GIL held (even though we actually do proper borrow checking now).

While I agree that our usage of `Send` is not really sound, I don't fixing this now and without auto traits is worth it. Especially so as using `send_wrapper` is usually a big warning sign that something strange is going. Meaning that I don't think this will happen accidentally.

On the other hand, pushing currently and eventually safe usage towards unsafe code will make this more difficult for every user which I suspect will mean more errors in aggregate compared to us being unable to close the `send_wrapper` loophole.

Reply:
Fair enough, let's leave `allow_threads` unchanged for now. I've added #2141 to the 1.0 project to remind us to see if we can work out a better solution over time.

Reply:
I added more explicit language concerning the soundness loophole and its fix. I also included the converse of the `Rc`-`compile_fail` doctest in the nightly version.